thunder/executors/fa3ex.py

```diff
@@ -0,0 +1,156 @@
+import torch
+
+# flash attn 3
+try:
+    from flash_attn_interface import _flash_attn_forward, _flash_attn_backward, flash_attn_func
+
+    HAS_FA3 = True
+except:
+    HAS_FA3 = False
+
+import thunder
+from thunder.core.transforms import get_grad, put_grads
+from thunder.extend import OperatorExecutor, register_executor
+
+fa3_ex: OperatorExecutor = OperatorExecutor("fa3", version="0.1")
+register_executor(fa3_ex)
+
+
+def fa3_fwd_meta(
+    q: thunder.torch.TensorLike,
+    k: thunder.torch.TensorLike,
+    v: thunder.torch.TensorLike,
+    causal: bool = False,
+    softmax_scale: float | None = None,
+):
+    return thunder.TensorProxy(like=q), thunder.TensorProxy(like=q, shape=(*q.shape[:1], 1))
+
+
+def fa3_fwd_impl(
+    q: torch.Tensor, k: torch.Tensor, v: torch.Tensor, causal: bool = False, softmax_scale: float | None = None
+):
+    if not HAS_FA3:
+        raise Exception("fa3 not built, cannot use fa3 executor")  # checker should fail before getting here
+
+    if softmax_scale is None:
+        softmax_scale = q.shape[-1] ** (-0.5)
+    out, q, k, v, out_padded, softmax_lse, S_dmask = _flash_attn_forward(q, k, v, softmax_scale, causal)
+    return out, softmax_lse
+
+
+def fa3_bwd_meta(
+    dout: thunder.torch.TensorLike,
+    q: thunder.torch.TensorLike,
+    k: thunder.torch.TensorLike,
+    v: thunder.torch.TensorLike,
+    out: thunder.torch.TensorLike,
+    softmax_lse,
+    causal: bool = False,
+    softmax_scale: None | float = None,
+):
+
+    grads = (thunder.TensorProxy(like=q), thunder.TensorProxy(like=k), thunder.TensorProxy(like=v))
+    return grads
+
+
+def fa3_bwd_impl(
+    dout: torch.Tensor,
+    q: torch.Tensor,
+    k: torch.Tensor,
+    v: torch.Tensor,
+    out: torch.Tensor,
+    softmax_lse,
+    causal: bool = False,
+    softmax_scale: None | float = None,
+):
+    if not HAS_FA3:
+        raise Exception("fa3 not built, cannot use fa3 executor")  # checker should fail before getting here
+
+    dq, dk, dv = torch.empty_like(q), torch.empty_like(k), torch.empty_like(v)
+    if softmax_scale is None:
+        softmax_scale = q.shape[-1] ** (-0.5)
+    _flash_attn_backward(
+        dout,
+        q,
+        k,
+        v,
+        out,
+        softmax_lse,
+        dq,
+        dk,
+        dv,
+        softmax_scale,
+        causal,
+    )
+    dq = dq[..., : dout.shape[-1]]  # We could have padded the head dimension
+    dk = dk[..., : dout.shape[-1]]
+    dv = dv[..., : dout.shape[-1]]
+    grads = (dq, dk, dv)
+    return grads
+
+
+fa3_fwd = fa3_ex.register_operator("fa3_fwd", meta=fa3_fwd_meta, fn=fa3_fwd_impl)
+fa3_bwd = fa3_ex.register_operator("fa3_bwd", meta=fa3_bwd_meta, fn=fa3_bwd_impl)
+
+
+def fa3_checker(query, key, value, attn_mask=None, dropout_p=0.0, is_causal=False, scale=None):
+    if not HAS_FA3:
```

Comment:
Does `fa3` support a specific cuda version?

`device_version_support` is a helper function in `thunder/tests/test_sdpaex_executor.py`. You should move it to a more general place.

```python
    # Flash Attention 3 is only supported on Hopper devices.
    if not device_version_support(query.device.type, CudaVersion(9, 0)):
        return False
```

Reply:
Under the requirements in this section https://github.com/Dao-AILab/flash-attention?tab=readme-ov-file#flashattention-3-beta-release, it mentions requires cuda >= 12.3. Do we want to add an additional check for sm >= 90? I think the current HAS_FA3 check is sufficient (see the below response to @mruberry's comment) and I was planning to just add some clarifying comments in the code explaining it, but I can also add another check if it would make things clearer.

Reply:
Added clarifying comment